File: solutions/javascript/perfect-numbers/1/perfect-numbers.js
class PerfectNumbers {
    classify(input) {
        if (input < 1) { return 'Classification is only possible for natural numbers.'; } // just going by tests, but would also check it's an integer...
        const aliquot = this.aliquot(input);
        if (aliquot === input) { return 'perfect'; }
        if (aliquot > input) { return 'abundant'; }
        return 'deficient';
    }
    aliquot(input) {
        let aliquot = 0;
		for (let possibleFactor = 1; possibleFactor < input; possibleFactor++) {
			if (input % possibleFactor === 0) {
				aliquot += possibleFactor;
			}
		}
        return aliquot;
    }
}
export default PerfectNumbers;

File: solutions/elixir/perfect-numbers/1/lib/perfect_numbers.ex
defmodule PerfectNumbers do
  @doc """
  Determine the aliquot sum of the given `number`, by summing all the factors
  of `number`, aside from `number` itself.

  Based on this sum, classify the number as:

  :perfect if the aliquot sum is equal to `number`
  :abundant if the aliquot sum is greater than `number`
  :deficient if the aliquot sum is less than `number`
  """
  @spec classify(number :: integer) :: {:ok, atom} | {:error, String.t()}
  def classify(number) when number < 1, do: {:error, "Classification is only possible for natural numbers."}
  def classify(number) do
    case 1..(number-1)//1 |> Enum.filter(&(rem(number, &1) == 0)) |> Enum.sum() do
      ^number -> {:ok, :perfect}
      x when x > number -> {:ok, :abundant}
      _ -> {:ok, :deficient}
    end
  end
end
